src/libs/ajax/GoogleStorage.ts

```diff
@@ -1,12 +1,77 @@
 import * as Utils from 'src/libs/utils'
 import { getExtension, getFileName, tools } from 'src/pages/workspaces/workspace/analysis/notebook-utils'
 
 
+/*
+ * Detects errors due to requester pays buckets, and adds the current workspace's billing
+ * project if the user has access, retrying the request once if necessary.
+ */
+const withRequesterPays = wrappedFetch => (url, ...args) => {
+  const bucket = /\/b\/([^/?]+)[/?]/.exec(url)![1]
+  const workspace = workspaceStore.get()
+
+  const getUserProject = async () => {
+    if (!requesterPaysProjectStore.get() && workspace && await canUseWorkspaceProject(workspace)) {
+      requesterPaysProjectStore.set(workspace.workspace.googleProject)
+    }
+    return requesterPaysProjectStore.get()
+  }
+
+  const tryRequest = async () => {
+    const knownRequesterPays = knownBucketRequesterPaysStatuses.get()[bucket]
+    try {
+      const userProject = (knownRequesterPays && await getUserProject()) || undefined
+      const res = await wrappedFetch(Utils.mergeQueryParams({ userProject }, url), ...args)
+      !knownRequesterPays && knownBucketRequesterPaysStatuses.update(_.set(bucket, false))
+      return res
+    } catch (error) {
+      if (knownRequesterPays === false) {
+        throw error
+      } else {
+        const newResponse = await checkRequesterPaysError(error)
+        if (newResponse.requesterPaysError && !knownRequesterPays) {
+          knownBucketRequesterPaysStatuses.update(_.set(bucket, true))
+          if (await getUserProject()) {
+            return tryRequest()
+          }
+        }
+        throw newResponse
+      }
+    }
+  }
+  return tryRequest()
+}
+
+// requesterPaysError may be set on responses from requests to the GCS API that are wrapped in withRequesterPays.
+// requesterPaysError is true if the request requires a user project for billing the request to. Such errors
+// are not transient and the request should not be retried.
+export const fetchBuckets = _.flow(withRequesterPays, withRetryOnError(error => Boolean(error.requesterPaysError)), withUrlPrefix('https://storage.googleapis.com/'))(fetchOk)
+
+/**
+ * Only use this if the user has write access to the workspace to avoid proliferation of service accounts in projects containing public workspaces.
+ * If we want to fetch a SA token for read access, we must use a "default" SA instead (api/google/user/petServiceAccount/token).
+ */
+const getServiceAccountToken = Utils.memoizeAsync(async (googleProject, token) => {
+  const scopes = ['https://www.googleapis.com/auth/devstorage.full_control']
+  const res = await fetchSam(
+    `api/google/v1/user/petServiceAccount/${googleProject}/token`,
+    _.mergeAll([authOpts(token), jsonBody(scopes), { method: 'POST' }])
+  )
+  return res.json()
+}, {
+  expires: 1000 * 60 * 30,
+  // @ts-expect-error
+  keyFn: (...args) => JSON.stringify(args)
+})
+
+export const saToken = googleProject => getServiceAccountToken(googleProject, getUser().token)
```

Comment:
consider wrapping this in `GoogleStorage` module

Reply:
I'm not sure about this one. All the methods of GoogleStorage use the GCS API, whereas this one uses Sam. I could see adding this as a method of Ajax, but I think long term we will probably be moving away from funneling everything through Ajax and using more direct imports of the submodules (GoogleStorage, etc).

Reply:
Yeah, I'm fine with not funneling through Ajax, but I like the idea of all API calls being wrapped in a module where the expectation is that each method makes an api call. Good compromise could be to add a type signature that indicates this returns a Promise.

Reply:
Good point. Definitely should take advantage of this being in a TS module now to add types.